docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,127 @@
+SDK.
+
+## EDL installation
+
+In order to support ocall opt-in, system EDL files will now be installed as part of the SDK at
+`${install_path}/include/openenclave/edl/`. This allows them to be imported into user edl by the
+enclave developer.
+
+## Developer opt-in
+
+In order to use any system OCall, it must be imported into the enclave's EDL file. If it is not
+included, but the OCall is still called by the enclave application or some part of the SDK,
+enclave linking will fail.
+
+Users can import a system OCall the same way any other edl file would be imported. For example,
+if an enclave with edl file `sample.edl` wants to use the function `oe_log_ocall()` they can do
+one of the following:
+
+*Note: These examples assume that `${install_path}/include` is passed to edger8r as an edl search path*
+
+1. Import just the required OCall
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import oe_log_ocall;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+2. Import all OCalls from `tee.edl`
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import *;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+In example (1), linking would fail if the enclave made a call to `oe_realloc_ocall()`
+(also part of tee.edl) unless the developer modified the import line to
+`from "openenclave/edl/tee.edl" import oe_log_ocall, oe_realloc_ocall;`. In this scenario,
+example (2) will work without modification.
+
+## EDL local structures
+
+Under this design, system edl files will prefer using local structures; that is, structures
+that are defined in edl. This is distinct from foreign structures, which are defined in
+C/C++ header files and included in the edl file.
+
+In the case that a foreign structure must be used, system edl files may only include headers
+that are installed with the SDK.
+
+Local EDL structures provide additional security benefits detailed [here](
+https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/full_edl_serialization.md#motivation)
+so foreign structures should be avoided whenever possible.
+
+## Developer opt-out
+
+Because system OCalls are generally called by the SDK implementation, developers
+may not have direct control over which OCalls will be called by their application.
+A developer can refuse to import a function in EDL which will cause a linker error
+if it is used, but they should also be able to take action to resolve this failure.
+
+To this end, any OE SDK feature which makes an OCall on behalf of an enclave
+should provide a mechanism to disable the OCall at compile-time and fail
+gracefully.
+
+## Exceptions to edl import and opt-in/out mechanisms
+
+A small number of OCalls in OE SDK are not implemented in edl and cannot be for
```

Comment:
This is no longer true.
With #2551, a context parameter is passed to ecalls. ecall_context_t contains enough space (16 KB) to make all the core ocalls (and most user written ocalls too) without any additional ocalls for host memory allocation/deallocation.

With the above guarantee provided by the EDL mechanism,   malloc/free ocalls can be themselves be implemented in EDL since their tiny parameter size (maximum declared parameter size = sizeof(pointer) + sizeof(uint64_t) = 16 bytes) guarantees that they can be dispatched without any additional ocalls for host memory allocation.

We are pretty much at a point where *all* ocalls can be EDL generated.

Reply:
That's awesome!

Reply:
We should update this doc to that effect and enumerate these as tasks rather than exceptions then.

As an extension of this, can we make a stronger statement about the developer opt-out? i.e. The design prescription is that OE SDK ocalls must be expressed as EDLs and the opt-out mechanism is always to not include the EDL, which renders this paragraph irrelevant, no?

> To this end, any OE SDK feature which makes an OCall on behalf of an enclave should provide a mechanism to disable the OCall at compile-time and fail gracefully.

The one remaining exception to this seems to be cpuid, and that's a case where there simply is no opt-out and no proposal for a mechanism to disable it with a graceful fallback mechanism. You would have to rewrite the initialization to pass the table in on initialization rather than have it call out to avoid this, and maybe that's a strategy that should be documented here.

Reply:
Good point about cpuid. With @jhand2 recent PR (215779f) to allow exception handler to make ocalls, cpuid table could be fetched  from host lazily only when a cpuid instruction is executed by an enclave.